### PR TITLE
Harden type checking for enum values

### PR DESCRIPTION
**Pull request by liufengyun**
This is a follow-up of #11327.

We cannot generalize further without complications:

```Scala
val b = 'b'

56 match
case 'a' =>
case `b` =>
```

### CODE REVIEW

**Comment by liufengyun** [comment]
The change detects a possible error in the following code:

https://github.com/lampepfl/dotty/blob/05e5eabfd613f1ad72ee760c112f6a13813b15f5/scaladoc/src/dotty/tools/scaladoc/translators/ScalaSignatureProvider.scala#L25-L26

```scala
Error:  -- [E007] Type Mismatch Error: /__w/dotty/dotty/scaladoc/src/dotty/tools/scaladoc/translators/ScalaSignatureProvider.scala:25:27 
Error:  25 |      case Kind.Given(Kind.Val, _, _) =>
Error:     |                      ^^^^^^^^
Error:     |Found:    (dotty.tools.scaladoc.Kind.Val : dotty.tools.scaladoc.Kind)
Error:     |Required: dotty.tools.scaladoc.Kind.Def | dotty.tools.scaladoc.Kind.Class
Error:     |pattern type is incompatible with expected type
```

File: tests/run/option-extract.scala [hunk 1]
```diff
@@ -6,7 +6,7 @@ enum Option[+A]:
   opaque type ExtractResult[B] = (=> B) => B
 
   def extract[B](f: A => B): ExtractResult[B] =
-    def result(default: => B): B = this match
+    def result(default: => B): B = (this: Option[A]) match
```
Comment (smarter) [inline tests/run/option-extract.scala:9]: Why is this change needed?

Comment (michelou) [reply tests/run/option-extract.scala:9]: @liufengyun Here, I was asking myself the same question !

Comment (liufengyun) [reply tests/run/option-extract.scala:9]: Good question. We need to widen the scrutinee type instead.

Comment (liufengyun) [reply tests/run/option-extract.scala:9]: Actually, we have the following error without the change:

```Scala
-- [E007] Type Mismatch Error: tests/run/option-extract.scala:10:11 ----------------------------------------------------
10 |      case None => default
   |           ^^^^
   |           Found:    (Option.None : Option[Nothing])
   |           Required: Any{ExtractResult = LazyRef([B] =>> (=> B) => B)} & Option[A]
   |           pattern type is incompatible with expected type
```

This seems to be a subtle interaction with opaque types in subtype checking.

Comment (liufengyun) [reply tests/run/option-extract.scala:9]: I tried to play with the opaque types with the following change:

``` diff
--- a/compiler/src/dotty/tools/dotc/core/Types.scala
+++ b/compiler/src/dotty/tools/dotc/core/Types.scala
@@ -4548,6 +4548,7 @@ object Types {

     private var selfTypeCache: Type = null
     private var appliedRefCache: Type = null
+    private var selfTypeWithoutOpaqueCache: Type = null

     /** The self type of a class is the conjunction of
      *   - the explicit self type if given (or the info of a given self symbol), and
@@ -4566,6 +4567,23 @@ object Types {
       selfTypeCache
     }

+    /** The self type without opaque members refinements */
+    def selfTypeWithoutOpaque(using Context): Type = {
+      def dropRefinement(tp: Type): Type = tp match
+        case RefinedType(parent, _, _: TypeAlias) =>
+          // TODO: maybe strengthen the check for Opaque flag
+          parent
+        case AndType(tp1, tp2) =>
+          dropRefinement(tp1) & dropRefinement(tp2)
+        case _ =>
+          tp
+      end dropRefinement
+
+      if selfTypeWithoutOpaqueCache == null then
+        selfTypeWithoutOpaqueCache = dropRefinement(selfType)
+      selfTypeWithoutOpaqueCache
+    }
```

But found no place where we could sensibly use `selfTypeWithoutOpaque` in the code.

Comment (liufengyun) [reply tests/run/option-extract.scala]: I managed to do this: in widen we always strip the opaque refinement for `ThisType`. See bf3b50a

Comment (smarter) [reply tests/run/option-extract.scala]: I'm not sure if stripping these refinements by default is a good idea since it's not clear in what situation this is safe to do. Also this isn't specific to opaque refinements:
```scala
trait Mixin

sealed trait Foo { self: Mixin =>
  def foo = this match {
    case this =>
  }
}
```
```scala
-- [E029] Pattern Match Exhaustivity Warning: try/selfmatch.scala:4:12 ----------------------------------------------------------------------------------------------------------------------------------------------------------------------
4 |  def foo = this match {
  |            ^^^^
  |            match may not be exhaustive.
  |
  |            It would fail on pattern case: _: Mixin & Foo
```

I think the exhaustiveness checking logic should just ignore self-types in general.

Comment (liufengyun) [reply tests/run/option-extract.scala]: This PR follows #11327, which is part of type checking, not exhaustivity check. It issues errors, instead of warnings. 

The motivation is to detect more bugs in programs: the error found in the compiler in this PR is evidence of its usefulness.

The example you posted above is another orthogonal issue --- it comes from the fact that we always widen the scrutinee in exhaustivity check.

> I'm not sure if stripping these refinements by default is a good idea since it's not clear in what situation this is safe to do.

Currently, the refinemen is stripped in widening, and kept in member selection. What cases you have mind that this PR will be a problem?

We have several choice here:

- Close this PR and revert #11327 for consistency
- Keep this PR w/o stripping and change the test case `tests/run/option-extract.scala` with `(this: Option[A])`
- Keep this PR with stripping

Comment (smarter) [reply tests/run/option-extract.scala]: > Currently, the refinemen is stripped in widening, and kept in member selection. What cases you have mind that this PR will be a problem?

I don't know, but I see things the other way around: we shouldn't do more stripping than we need to, and we shouldn't special-case opaque types more than we have to.

> The example you posted above is another orthogonal issue --- it comes from the fact that we always widen the scrutinee in exhaustivity check.

I believe the following should also compile without warnings:
```scala
trait Mixin

sealed trait Foo { self: Mixin =>
  def foo = this match {
    case _: Foo =>
  }
}
```
It really seems that the cleanest fix for this sort of issues is for the exhaustiveness checker to always strip self types.

Comment (liufengyun) [reply tests/run/option-extract.scala]: I kind of agree that stripping of opaque refinements should be avoided if possible. Among the following 3 choices:

1. Close this PR and revert #11327 for consistency
1. Keep this PR w/o stripping and change the test case `tests/run/option-extract.scala` with `(this: Option[A])`
1. Keep this PR with stripping

Maybe the 2nd is a good compromise?

BTW, this PR and `#11327` are not fixing any existing issues. They intend to harden type checking and align with Scala 2 behavoir in type checking patterns.

Comment (smarter) [reply tests/run/option-extract.scala]: > Maybe the 2nd is a good compromise?

I'm fine with that if we also open an issue to keep track of the issue with the exhaustivity checker.

Comment (liufengyun) [reply tests/run/option-extract.scala]: > I'm fine with that if we also open an issue to keep track of the issue with the exhaustivity checker.

Agreed. I'll create an issue for the test `tests/run/option-extract.scala` (it's a typing error, not an exhaustivity warning).

Comment (liufengyun) [reply tests/run/option-extract.scala:9]: The last commit removed and we created an issue #11669 for `tests/run/option-extract.scala`.

**Review by smarter — APPROVED** [review]
